content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,276 @@
+
+## Installation
+[FFTW](https://www.fftw.org) is one of the requirements of pyFFTW and it does not come with pyFFTW. You can install it through `conda`. I assume you have `conda` installed, and you know how to create and activate a conda environment. See [pyFFTw's github page](https://github.com/pyFFTW/pyFFTW) to learn more about pyFFTW installation, including the minimum Python version requirement.
+
+Once you are in your environment, install `FFTW` by running the following command:
+```bash
+conda install conda-forge::fftw
+```
+
+And, the other dependencies can be installed when installing pyFFTW:
+
+```bash
+conda install conda-forge::pyfftw
+``` 
+
+
+Now, we are ready to use pyFFTW! The following code demonstrates how to use pyFFTW to perform RFFT on a real-valued array. 
+
+```python
+import numpy as np
+import pyfftw
+
+T = np.random.rand(4)
+rfft_obj = pyfftw.builders.rfft(np.empty(len(T), dtype='float64'))
+R = rfft_obj(T)
+```
+
+## Performance Optimization
+RFFT performance varies across input sizes. The performance of RFFT can be significantly improved when input arrays have specific sizes (e.g., powers of two). Here we consider input sizes that are powers of two, and try to find ways to improve the performance of RFFT further in those cases. 
+
+
+### Timing
+Let's start by preparing a script that can get the timing of RFFT computation.
+
+```python
+def _get_timing_single(n, rfft_caller, timeout=5.0, iter_max=100000):
+    """
+    For a single input size n, return the time taken to compute RFFT using rfft_caller
+    """
+    pyfftw.forget_wisdom()  # clean up any previous plans
+    T = np.random.rand(n)
+
+    R = rfft_caller(T) # dummy to create wisdom
+    np.testing.assert_allclose(R, np.fft.rfft(T))  # verify correctness
+
+    total_time = 0.0
+    count = 0
+    while total_time < timeout and count < iter_max:
+        start_time = time.perf_counter()
+        rfft_caller(T)
+        total_time += time.perf_counter() - start_time
+        count += 1
+
+    return total_time / count
+
+
+def get_timing(n_values, rfft_caller, timeout=5.0, iter_max=1000000, verbose=True):
+    """
+    n_values: an array of input sizes
+    """
+    timing = np.full(len(n_values), -1.0, dtype='float64')
+    for i, n in enumerate(n_values):
+        timing[i] = _get_timing_single(n, rfft_caller, timeout=timeout, iter_max=iter_max)
+        if verbose:
+            print(f"log2(n) --> {int(np.log2(n))}", flush=True)
+    return timing
+```
+
+There are a few things to note in the above code:
+1. To be on the safe side, we use `pyfftw.forget_wisdom()` to clear all previously-stored plans (wisdoms). This is to make sure that we start fresh and our timing result does not benefit, in any way, from any previously-computed wisdom. If you are not familiar with the concept of wisdom in FFTW, you can read more about it [on pyFFTW's documentation](https://pyfftw.readthedocs.io/en/latest/source/pyfftw/pyfftw.html#wisdom-functions).
```

Comment:
You should consider providing more information about what you learned about "wisdom" and maybe it can be a separate (shorter) blog post (like a nice note to yourself) that gets cross referenced here. I can guarantee you that people will stop reading this post because you're glossing over wisdom. Either it's important or it's not!